packages/remix-dev/vite/plugin.ts

```diff
@@ -927,22 +929,88 @@ export const remixVitePlugin: RemixVitePlugin = (options = {}) => {
       },
     },
     {
-      name: "remix-empty-server-modules",
+      name: "remix-dot-server",
       enforce: "pre",
-      async transform(_code, id, options) {
+      async resolveId(id, importer, options) {
         if (options?.ssr) return;
+
+        let isResolving = options?.custom?.["remix-dot-server"] ?? false;
+        if (isResolving) return;
+        options.custom = { ...options.custom, "remix-dot-server": true };
+        let resolved = await this.resolve(id, importer, options);
+        if (!resolved) return;
+
         let serverFileRE = /\.server(\.[cm]?[jt]sx?)?$/;
         let serverDirRE = /\/\.server\//;
-        if (serverFileRE.test(id) || serverDirRE.test(id)) {
-          return {
-            code: "export {}",
-            map: null,
-          };
+        let isDotServer =
+          serverFileRE.test(resolved!.id) || serverDirRE.test(resolved!.id);
+        if (!isDotServer) return;
+
+        if (!importer) throw Error(`Importer not found: ${id}`);
+
+        let pluginConfig = await resolvePluginConfig();
+        let importerShort = path.relative(pluginConfig.rootDirectory, importer);
+        let isRoute = getRoute(pluginConfig, importer);
+
+        if (isRoute) {
+          let serverOnlyExports = SERVER_ONLY_EXPORTS.map(
+            (xport) => `\`${xport}\``
+          ).join(", ");
+          throw Error(
+            [
+              colors.red(`Server-only module referenced by client`),
+              "",
+              `    '${id}' imported by route '${importerShort}'`,
+              "",
+              `  The only route exports that can reference server-only modules are: ${serverOnlyExports}`,
+              `  but other route exports in '${importerShort}' depend on '${id}'.`,
+              "",
+              "  For more see https://remix.run/docs/en/main/discussion/server-vs-client",
+              "",
+            ].join("\n")
+          );
         }
+
+        let importedBy = path.parse(importerShort);
+        let ext = importedBy.ext === ".jsx" ? ".js" : ".ts";
+        let dotServerDir = path
+          .join(
+            path.basename(pluginConfig.appDirectory),
+            ".server",
+            "utils" + ext
+          )
+          .replace(/\.jsx$/, ".js");
+        let dotServerFile = path.join(
+          importedBy.dir,
+          importedBy.name + ".server" + ext
+        );
+
+        throw Error(
+          [
+            colors.red(`Server-only module referenced by client`),
+            "",
+            `    '${id}' imported by '${importerShort}'`,
+            "",
+            `  * If all code in '${importerShort}' is server-only:`,
+            "",
+            `    Rename it to '${dotServerFile}'`,
```

Comment:
I'm wondering if this should also mention `.server` directories?

Reply:
This is meant to be a "quickfix" option to help unblock anyone migrating. We could maybe reword to make that clearer. That said, I can't think of a good way to mention `.server` directories that doesn't feel heavier. Any suggestions?